File: test/unit/register_prompts_test.go
package unit

import (
	"embed"
	"encoding/json"
	"strings"
	"testing"

	"github.com/harness/harness-mcp/cmd/harness-mcp-server/config"
	"github.com/harness/harness-mcp/pkg/harness/prompts"
	p "github.com/harness/harness-mcp/pkg/prompts"
	"github.com/mark3labs/mcp-go/server"
)

func registerPromptsCopy(moduleID string, cfg *config.Config, mcpServer *server.MCPServer) (*p.Prompt, error) {
	
	    // Create a map to store prompts by mode
		modulePromptsByMode := map[string][]prompts.PromptFile{
			string(p.Standard): {},
			string(p.Architect): {},
		}
		
		// Get module-specific prompts for standard mode
		standardPrompts, err := mockGetModulePrompts(prompts.PromptFiles, strings.ToLower(moduleID), cfg.Internal, string(p.Standard))
		if err != nil {
			return nil, err
		}
		modulePromptsByMode[string(p.Standard)] = standardPrompts
		
		// Get module-specific prompts for architect mode (only for internal)
		if cfg.Internal {
			architectPrompts, err := mockGetModulePrompts(prompts.PromptFiles, strings.ToLower(moduleID), cfg.Internal, string(p.Architect))
			if err != nil {
				return nil, err
			}
			modulePromptsByMode[string(p.Architect)] = architectPrompts
		}
		
		// Check if we have any prompts for this module
		totalPrompts := len(modulePromptsByMode[string(p.Standard)]) + len(modulePromptsByMode[string(p.Architect)])
		if totalPrompts == 0 {
			// No prompts for this module
			return nil, nil
		}
	
		// Create a map to store combined content for each mode
		modeContents := make(map[string]string)
		
		// Get description and result description from the first available prompt
		description := ""
		resultDescription := ""
		
		// First, prioritize standard mode for descriptions to ensure deterministic behavior
		if len(modulePromptsByMode[string(p.Standard)]) > 0 {
			standardPrompts := modulePromptsByMode[string(p.Standard)]
			if standardPrompts[0].Metadata.Description != "" {
				description = standardPrompts[0].Metadata.Description
			}
			if standardPrompts[0].Metadata.ResultDescription != "" {
				resultDescription = standardPrompts[0].Metadata.ResultDescription
			}
		}
		
		// If standard mode doesn't have descriptions, fall back to architect mode
		if description == "" || resultDescription == "" {
			if len(modulePromptsByMode[string(p.Architect)]) > 0 {
				architectPrompts := modulePromptsByMode[string(p.Architect)]
				if description == "" && architectPrompts[0].Metadata.Description != "" {
					description = architectPrompts[0].Metadata.Description
				}
				if resultDescription == "" && architectPrompts[0].Metadata.ResultDescription != "" {
					resultDescription = architectPrompts[0].Metadata.ResultDescription
				}
			}
		}
		
		// Process each mode separately to build the content map
		for mode, modePrompts := range modulePromptsByMode {
			if len(modePrompts) == 0 {
				continue // Skip empty modes
			}
			
			// Combine all prompt contents for this mode
			var combinedContent strings.Builder
			for i, promptFile := range modePrompts {
				if promptFile.Content != "" {
					if i > 0 {
						combinedContent.WriteString("\n\n")
					}
					combinedContent.WriteString(promptFile.Content)
				}
			}
			
			// Store the combined content for this mode
			modeContents[mode] = combinedContent.String()
		}
		
		// Convert the mode contents map to JSON
		contentJSON, err := json.Marshal(modeContents)
		if err != nil {
			return nil, err
		}
		
		// Create a single MCP prompt with the module ID as the name
		mcpPrompt := p.NewPrompt().
			SetName(strings.ToUpper(moduleID)). // Use moduleID as the prompt name
			SetDescription(description).
			SetResultDescription(resultDescription).
			SetText(string(contentJSON)). // Store the JSON map as the prompt text
			Build()
		
		// Create a Prompts collection with the single prompt
		mcpPrompts := p.Prompts{}
		mcpPrompts.Append(mcpPrompt)
		
		// Register the prompt with the MCP server
		p.AddPrompts(mcpPrompts, mcpServer)
		
		return mcpPrompt, nil
}

// Mock the GetModulePrompts function to return our test prompts
func mockGetModulePrompts(fs embed.FS, module string, isInternal bool, mode string) ([]prompts.PromptFile, error) {
	// Return different prompts based on the requested mode
	if mode == string(p.Standard) {
		return []prompts.PromptFile{
			{
				Metadata: prompts.PromptMetadata{
					Description:       "Standard test description",
					ResultDescription: "Standard result description",
					Module:            "testmodule",
				},
				Content:  "Standard content",
				FilePath: "files/internal/standard/testmodule.txt",
			},
		}, nil
	} else if mode == string(p.Architect) {
		return []prompts.PromptFile{
			{
				Metadata: prompts.PromptMetadata{
					Description:       "Architect test description",
					ResultDescription: "Architect result description",
					Module:            "testmodule",
				},
				Content:  "Architect content",
				FilePath: "files/internal/architect/testmodule.txt",
			},
		}, nil
	}

	return []prompts.PromptFile{}, nil
}

func TestRegisterPrompts(t *testing.T) {
	// We can't directly replace the GetModulePrompts function, so we'll use our own implementation
	// in the testRegisterPrompts function

	cfg := &config.Config{
		Internal: true,
	}

	// Create fake MCP server
	mcpServer := server.NewMCPServer("test-server", "test-version")

	// Use our test version of registerPrompts
	prompt, err := registerPromptsCopy("testmodule", cfg, mcpServer)
	if err != nil {
		t.Fatalf("registerPrompts returned error: %v", err)
	}
	if prompt.Name != "TESTMODULE" {
		t.Errorf("expected prompt name 'TESTMODULE', got %q", prompt.Name)
	}
	if !strings.Contains(prompt.Description, "Standard test description") {
		t.Errorf("description mismatch, got %q", prompt.Description)
	}
	if !strings.Contains(prompt.ResultDescription, "Standard result description") {
		t.Errorf("result description mismatch, got %q", prompt.ResultDescription)
	}

	// Verify JSON text contains both modes
	var modeContents map[string]string
	if err := json.Unmarshal([]byte(prompt.Text), &modeContents); err != nil {
		t.Fatalf("failed to unmarshal prompt text: %v", err)
	}

	if modeContents[string(p.Standard)] != "Standard content" {
		t.Errorf("standard mode content mismatch, got %q", modeContents[string(p.Standard)])
	}
	if modeContents[string(p.Architect)] != "Architect content" {
		t.Errorf("architect mode content mismatch, got %q", modeContents[string(p.Architect)])
	}
}

File: tests/unit/register_prompts_test.rs
use crate::config::Config;
use crate::harness::prompts::{PromptFile, PromptMetadata};
use crate::prompts::{Mode, Prompt, Prompts};
use serde_json;
use std::collections::HashMap;

/// Mock function to simulate GetModulePrompts
fn mock_get_module_prompts(module: &str, is_internal: bool, mode: &str) -> Result<Vec<PromptFile>, Box<dyn std::error::Error>> {
    match mode {
        "standard" => Ok(vec![PromptFile {
            metadata: PromptMetadata {
                description: "Testmodule prompt description".to_string(),
                result_description: "Testmodule prompt result description".to_string(),
                module: "testmodule".to_string(),
            },
            content: "Standard content".to_string(),
            file_path: "files/internal/standard/testmodule.txt".to_string(),
        }]),
        "architect" => Ok(vec![PromptFile {
            metadata: PromptMetadata {
                description: "Testmodule prompt description".to_string(),
                result_description: "Testmodule prompt result description".to_string(),
                module: "testmodule".to_string(),
            },
            content: "Architect content".to_string(),
            file_path: "files/internal/architect/testmodule.txt".to_string(),
        }]),
        _ => Ok(vec![]),
    }
}

/// Copy of register_prompts function for testing
fn register_prompts_copy(
    module_id: &str,
    cfg: &Config,
) -> Result<Option<Prompt>, Box<dyn std::error::Error>> {
    // Create a map to store prompts by mode
    let mut module_prompts_by_mode: HashMap<String, Vec<PromptFile>> = HashMap::new();
    module_prompts_by_mode.insert("standard".to_string(), vec![]);
    module_prompts_by_mode.insert("architect".to_string(), vec![]);

    // Get module-specific prompts for standard mode
    let standard_prompts = mock_get_module_prompts(
        &module_id.to_lowercase(),
        cfg.internal,
        "standard",
    )?;
    module_prompts_by_mode.insert("standard".to_string(), standard_prompts);

    // Get module-specific prompts for architect mode (only for internal)
    if cfg.internal {
        let architect_prompts = mock_get_module_prompts(
            &module_id.to_lowercase(),
            cfg.internal,
            "architect",
        )?;
        module_prompts_by_mode.insert("architect".to_string(), architect_prompts);
    }

    // Check if we have any prompts for this module
    let total_prompts: usize = module_prompts_by_mode.values().map(|v| v.len()).sum();
    if total_prompts == 0 {
        // No prompts for this module
        return Ok(None);
    }

    // Create a map to store combined content for each mode
    let mut mode_contents: HashMap<String, String> = HashMap::new();

    // Get description and result description from the first available prompt
    let mut description = String::new();
    let mut result_description = String::new();

    // Process each mode separately to build the content map
    for (mode, mode_prompts) in &module_prompts_by_mode {
        if mode_prompts.is_empty() {
            continue; // Skip empty modes
        }

        // Use the first prompt's metadata for description and result description if not already set
        if description.is_empty() && !mode_prompts[0].metadata.description.is_empty() {
            description = mode_prompts[0].metadata.description.clone();
        }

        if result_description.is_empty() && !mode_prompts[0].metadata.result_description.is_empty() {
            result_description = mode_prompts[0].metadata.result_description.clone();
        }

        // Combine all prompt contents for this mode
        let mut combined_content = Vec::new();
        for prompt_file in mode_prompts {
            if !prompt_file.content.is_empty() {
                combined_content.push(prompt_file.content.clone());
            }
        }

        // Store the combined content for this mode
        mode_contents.insert(mode.clone(), combined_content.join("\n\n"));
    }

    // Convert the mode contents map to JSON
    let content_json = serde_json::to_string(&mode_contents)?;

    // Create a single MCP prompt with the module ID as the name
    let mcp_prompt = Prompt::new()
        .set_name(&module_id.to_uppercase()) // Use moduleID as the prompt name
        .set_description(&description)
        .set_result_description(&result_description)
        .set_text(&content_json) // Store the JSON map as the prompt text
        .build();

    Ok(Some(mcp_prompt))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_register_prompts() {
        let cfg = Config {
            internal: true,
            ..Default::default()
        };

        // Use our test version of register_prompts
        let prompt = register_prompts_copy("testmodule", &cfg)
            .expect("register_prompts should not return error")
            .expect("register_prompts should return a prompt");

        assert_eq!(prompt.name, "TESTMODULE");
        assert!(prompt.description.contains("Testmodule prompt description"));
        assert!(prompt.result_description.contains("Testmodule prompt result description"));

        // Verify JSON text contains both modes
        let mode_contents: HashMap<String, String> = serde_json::from_str(&prompt.text)
            .expect("prompt text should be valid JSON");

        assert_eq!(
            mode_contents.get("standard").unwrap(),
            "Standard content"
        );
        assert_eq!(
            mode_contents.get("architect").unwrap(),
            "Architect content"
        );
    }

    #[test]
    fn test_register_prompts_external_config() {
        let cfg = Config {
            internal: false,
            ..Default::default()
        };

        let prompt = register_prompts_copy("testmodule", &cfg)
            .expect("register_prompts should not return error")
            .expect("register_prompts should return a prompt");

        // Verify JSON text contains only standard mode for external config
        let mode_contents: HashMap<String, String> = serde_json::from_str(&prompt.text)
            .expect("prompt text should be valid JSON");

        assert_eq!(
            mode_contents.get("standard").unwrap(),
            "Standard content"
        );
        assert!(!mode_contents.contains_key("architect"));
    }

    #[test]
    fn test_mock_get_module_prompts_standard() {
        let prompts = mock_get_module_prompts("testmodule", true, "standard")
            .expect("should return prompts");

        assert_eq!(prompts.len(), 1);
        assert_eq!(prompts[0].content, "Standard content");
        assert_eq!(prompts[0].metadata.module, "testmodule");
    }

    #[test]
    fn test_mock_get_module_prompts_architect() {
        let prompts = mock_get_module_prompts("testmodule", true, "architect")
            .expect("should return prompts");

        assert_eq!(prompts.len(), 1);
        assert_eq!(prompts[0].content, "Architect content");
        assert_eq!(prompts[0].metadata.module, "testmodule");
    }

    #[test]
    fn test_mock_get_module_prompts_unknown_mode() {
        let prompts = mock_get_module_prompts("testmodule", true, "unknown")
            .expect("should return empty prompts");

        assert_eq!(prompts.len(), 0);
    }
}